core/domain/stats_jobs_one_off.py

```diff
@@ -18,14 +18,466 @@
+            complete_exploration_events_for_this_version = [
+                val for val in complete_exploration_event_dicts if val[
+                    'version'] == version]
+            state_hit_events_for_this_version = [
+                val for val in state_hit_event_dicts if val[
+                    'version'] == version]
+            state_answer_events_for_this_version = [
+                val for val in state_answer_event_dicts if val[
+                    'version'] == version]
+            versioned_exploration = exp_services.get_exploration_by_id(
+                exp_id, version=version)
+
+            is_revert = False
+            if version == 1:
+                # Create default state stats mapping for the first version.
+                for state_name in versioned_exploration.states:
+                    state_stats_mapping[state_name] = (
+                        stats_domain.StateStats.create_default())
+            else:
+                exp_commit_log = exp_models.ExplorationCommitLogEntryModel.get(
+                    'exploration-%s-%s' % (exp_id, version))
+                change_list = exp_commit_log.commit_cmds
+
+                # Handling state additions, renames and deletions.
+                for change_dict in change_list:
+                    if change_dict['cmd'] == exp_domain.CMD_ADD_STATE:
+                        state_stats_mapping[change_dict['state_name']] = (
+                            stats_domain.StateStats.create_default())
+                    elif change_dict['cmd'] == exp_domain.CMD_DELETE_STATE:
+                        state_stats_mapping.pop(change_dict['state_name'])
+                    elif change_dict['cmd'] == exp_domain.CMD_RENAME_STATE:
+                        state_stats_mapping[change_dict['new_state_name']] = (
+                            state_stats_mapping.pop(
+                                change_dict['old_state_name']))
+                    elif change_dict['cmd'] == 'AUTO_revert_version_number':
+                        revert_to_version = change_dict['version_number']
+                        is_revert = True
+
+            # Computing num_starts and num_completions.
+            num_starts += len(start_exploration_events_for_this_version)
+            num_completions += len(complete_exploration_events_for_this_version)
+            # Dict mapping each session ID to the list of state hit events for
+            # that session.
+            state_hit_events_session_id_mapping = {}
+            # Dict mapping each state name of the exploration and the list of
+            # unique session IDs that have entered that state.
+            state_session_ids = {
+                state_name: [] for state_name in versioned_exploration.states
+            }
+            session_ids_set = set()
+            # Compute total_hit_count and first_hit_count for the states.
+            for state_hit_event in state_hit_events_for_this_version:
```

Comment:
Again, nested for loop; seems wasteful. Try to avoid.

Reply:
No, O(n) only measures asymptotic complexity. Two O(n) passes are still two passes; getting it down to one reduces the time spent by a factor of 2 (approximately; there's going to be some overhead too).

In this case, could you not compute the information on first pass through the state_hit_event list for all versions? E.g. store counts/session IDs in a nested dict with successive version number and state name keys? This might mean you'll need to get all state names for all versions of the exploration at the outset, but that's also reasonable since you're processing all versions in this step anyway.

Reply:
Done.